refactor(compiler-cli): add back ngcc as a no-op with a warning

leosvelperez opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

The Angular Compatibility Compiler is removed in Angular 16 and therefore, `ngcc` is no longer available in users' workspaces. In some setups, `ngcc` is invoked in the `postinstall` script of the `package.json`. When updating Angular, in scenarios where `postinstall` is run (e.g. Nx migrations), this throws an error before a migration can remove the `ngcc` invocation. This makes the update process to be more complicated.

## What is the new behavior?

`ngcc` is available as a no-op operation. When invoked, it will warn, providing details about removing `ngcc`. It gives some context at the tool level to folks who haven't caught up with news that `ngcc` was removed.

**Note:** In Angular 17, this will be removed.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I chose to inline some slightly modified functionality from chalk for the colors in the console. I did that to avoid adding an extra dependency for this. I'm open to adding the dependency instead or implementing a different way to get colors in the console in a cross-platform way if preferred.

JoostK commented on packages/compiler-cli/ngcc/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+
+ts_library(
+    name = "ngcc",
+    srcs = glob([
+        "*.ts",
+    ]),
```
We can avoid the `glob` here:

```suggestion
    srcs = ["index.ts"],
```

JoostK commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -0,0 +1,58 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+// https://github.com/chalk/chalk/blob/main/source/utilities.js#L21
```
Can these be stable URLs here and below, instead of using `main` which is a moving target (press `y` in Github to update the URL with the commit sha before copying).

JoostK reviewed (commented):
Thanks! Could you please change this to a `refactor` commit, as the current feat commit cannot be landed for 16.0-rc

dgp1130 commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -0,0 +1,58 @@
+
+==========================================
+
+ALERT: As of Angular 16, "ngcc" is no longer required and not invoked during CLI builds. You are seeing this message because the current operation invoked the "ngcc" command directly.
```
Not to relitigate our discussion in chat, but I just realized we aren't very clear that `ngcc` is safe to remove in v16. Maybe we can add:

> ALERT: As of Angular 16, "ngcc" is no longer required and not invoked during CLI builds. You are seeing this message because the current operation invoked the "ngcc" command directly. This "ngcc" invocation can be safely removed.

leosvelperez committed:
refactor(compiler-cli): add back ngcc as a no-op with a warning

This commit adds back `ngcc` as a no-op operation. When invoked it will warn providing details about removing `ngcc`.

In Angular 17, this will be removed.

JoostK reviewed (approved):
LGTM

JoostK commented:
This shouldn't have any impact on g3, but let's run a presubmit regardless to know there's no sync conflicts.

dgp1130 reviewed (approved):
Message looks good to me, I agree we shouldn't add a dependency for terminal coloring just for this so your solution looks fine.

pkozlowski-opensource commented:
caretaker note: none of the files in this PR impacts G3, no need to run presubmit

pkozlowski-opensource commented:
This PR was merged into the repository by commit cfab3ad706889cab3725a039bab20768b0f62fad.
